Configurable number of data nodes with scale up functionality

LLParse opened the pull request:
Datanodes now have a numeric ID in the task names like journalnodes and namenodes. A new 'ACTIVE' acquisition phase was introduced.

Similar to Myriad, the number of datanodes is configurable at runtime via REST API (only scale up for now). This newly-introduced configuration mutability presents a new challenge not yet accounted for. Now that in-memory configuration may deviate from file configuration, we will need to re-generate that file (mesos-site.xml) to prevent the cluster from reaching an inconsistent state if the scheduler were to restart. Alternatively, the configuration could be tracked in Zookeeper via persistent state. Whichever pattern we choose, we should try and follow it for any other config mutations at runtime.

Scale down functionality is not yet implemented, it is going to require a sizable amount of work to be done right and I would like to engage the community on its design. There are a few approaches we could take to ensure nodes are decommissioned and eventually stopped without risking data corruption or under-replicated blocks. Maybe this would be best as a new issue.

I refactored the JRE to be packaged in the distribution and served via the scheduler Jetty server instead of being downloaded from mesosphere's servers for each task. This made development without an internet connection feasible.

Also, in order to get valuable task states for data nodes, the infinite execution loop was removed from `hdfs-mesos-datanode`. This will also help to expose issues where a datanode is failing to start over and over, but the task still appears healthy indefinitely.

Fixes #112

Commit:
Configurable quantity of data nodes with scale up functionality

Commit:
Remove thread sleep

elingg commented on bin/hdfs-mesos-datanode:
```diff
@@ -2,24 +2,5 @@
-}
-
-run_datanode
+$DIR/hdfs-mesos-env
```
We need to loop through and run the datanode here in order to try to recover from failures. What if there is a network partition or a failure communicating with a NameNode or JournalNode? We don't want the task to die immediately, but we want to spend some time relaunching the task. This is why these scripts were written in the form of loops. (cc @brndnmtthws). I would also be okay writing java code that loops through and tries to launch a number of times. As another note, once we add health checks to our tasks, we will be able to kill a task if we aren't able to hit the http endpoint for datanode, namenode, etc

brndnmtthws replied:
I'd prefer if we kept the `while [ true ] do ; ... ; done` stuff for now.

Think of it as cheap and easy HA.

elingg replied:
+1

LLParse replied:
Works for me! I've added a `sleep 10` like the other nodes.

elingg commented on hdfs-commons/src/main/java/org/apache/mesos/hdfs/config/HdfsFrameworkConfig.java:
```diff
@@ -32,10 +32,14 @@
+  private static final int DEFAULT_DFS_REPLICATION = 3;
+  @SuppressWarnings("PMD")
+  private static final String DEFAULT_SERVER_HOST = "0.0.0.0";
 
```
If we are going to create one string as a constant here, we need to make all our strings constants. Perhaps we could create another constants file to separate out this logic. If not, we need to be consistent, and we shouldn't create constants for String values.

elingg commented on hdfs-commons/src/main/java/org/apache/mesos/hdfs/config/HdfsFrameworkConfig.java:
```diff
@@ -32,10 +32,14 @@
   private static final int DEFAULT_ZK_TIME_MS = 20000;
   private static final int DEFAULT_RECONCILIATION_TIMEOUT = 30;
   private static final int DEFAULT_DEADNODE_TIMEOUT = 90;
+  private static final int DEFAULT_DFS_REPLICATION = 3;
```
We shouldn't have to SuppressWarnings

elingg commented on hdfs-executor/src/main/java/org/apache/mesos/hdfs/executor/TaskSupervisor.java:
```diff
@@ -0,0 +1,52 @@
+package org.apache.mesos.hdfs.executor;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
```
I really like breaking this logic out, but if so we need to remove this logic from startProcess. We shouldn't have it in both places. It appears that it's trying to start the task twice.

LLParse replied:
See above comment

elingg commented on hdfs-executor/src/main/java/org/apache/mesos/hdfs/executor/NodeExecutor.java:
```diff
@@ -47,6 +47,7 @@ public void launchTask(final ExecutorDriver driver, final TaskInfo taskInfo) {
     startProcess(driver, task);
+    new Thread(new TaskSupervisor(driver, task)).start();
     driver.sendStatusUpdate(TaskStatus.newBuilder()
         .setTaskId(taskInfo.getTaskId())
```
If we're going to change this in NodeExecutor, we should also do this in NameNodeExecutor.

LLParse replied:
I'm reverting this logic. if we decide to overhaul the while loop HA pattern in the future, something similar should probably go in AbstractNodeExecutor to watch the process and update the task status accordingly (and re-run the task of course).

elingg commented on bin/hdfs-mesos-env:
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+
```
Thanks for breaking this logic out!

elingg commented on hdfs-scheduler/src/main/java/org/apache/mesos/hdfs/api/ClusterResource.java:
```diff
@@ -0,0 +1,64 @@
+public class ClusterResource {
+  private final Log log = LogFactory.getLog(ClusterResource.class);
+
+  private HdfsFrameworkConfig config;
```
If we're going to update the configuration while running, one solution is to update the mesos-site.xml file and copy the mesos-site.xml configuration file to the executors (similar to what we do for hdfs-site.xml). Otherwise we have an inconsistent state as you mentioned. In general, we need to clean up our configuration for this framework. This is a challenge since HDFS requires static configuration files.

LLParse replied:
Well, I'm not sure if we'd have to copy the mesos-site.xml file to the executors, since I don't think it actually gets used there (correct me if I'm wrong). However, the mesos-site.xml used by the scheduler would indeed need to be modified with the new value to ensure things go smoothly in a scheduler failover event.

I'd suggest we hold off on config-updating logic until we have a game plan for the configuration as a whole (with the understanding that the config might need to be adjusted if a scaled-up scheduler dies and needs restarting). Open to input, though.

elingg replied:
If we're going to update mesos-site.xml while running, we do make sure the changes get persisted to the appropriate places. This will also fix the problem that we need to configure mesos-site.xml at compile time not at run time for the changes to effect the executors. It is in use by executors. In fact the data directories the executors write to is specified there, i.e. mesos.hdfs.data.dir.

I completely understand that configuration is a separate problem, but this patch is putting configuration into an inconsistent state and if we do adjust the configuration while the code is running, we need to make sure it is persisted consistently for the scheduler and the executors. I'm worried about putting in a patch that will affect scheduler failover and the executors and will potentially create bugs that didn't previously exist.

elingg commented:
A major issue here is that we still need to distribute the binaries on slaves where the datanodes are not running. To do this we need to launch a simple task with few resources, which distributes the binaries there. If not, people running say Spark on HDFS will encounter errors because they cannot fetch files on the slaves where the binaries are not distributed.

elingg commented:
Thanks for the great work @LLParse! I listed my concerns that need to be resolved above. Also, the code needs to compile (it is failing the build). It also needs to be tested on Mesos and DCOS to make sure there are no regressions introduced.

elingg commented on hdfs-commons/src/main/java/org/apache/mesos/hdfs/config/HdfsFrameworkConfig.java:
```diff
@@ -257,6 +269,14 @@ public String getFrameworkHostAddress() {
   }
 
+  public String getConfigServerHost() {
+    String configServerHost = System.getProperty("mesos.hdfs.config.server.host");
```
We should be following the same format as the other config values here instead of using System.getProperty

LLParse replied:
I was originally following the getConfigServerPort() pattern with the system property override, but upon further inspection we don't need this at all. Updated connector provider accordingly.

elingg commented on hdfs-commons/src/main/java/org/apache/mesos/hdfs/config/HdfsFrameworkConfig.java:
```diff
@@ -275,6 +295,14 @@ public int getDeadNodeTimeout() {
   }
 
+  public int getDFSReplication() {
+    String dfsReplicationString = System.getProperty("dfs.replication");
```
We should be following the same format as the other config values here instead of using System.getProperty

LLParse replied:
Agreed.

elingg commented on hdfs-scheduler/src/main/java/org/apache/mesos/hdfs/api/ClusterResource.java:
```diff
@@ -0,0 +1,64 @@
+
+  @POST
+  @Path("scaledown")
+  public Response scaleDown(@QueryParam("instances") int instances) {
```
Can we remove scaleDown since we are implementing it yet? I don't think we need to implement it yet. Let's start with scaleUp

LLParse replied:
Agreed, removed.

brndnmtthws commented on bin/build-hdfs:
```diff
@@ -104,7 +114,7 @@ cd -
 echo "Compressing to $EXECUTOR.tgz"
 rm -f $BUILD_CACHE_DIR/$EXECUTOR.tgz
 cd $BUILD_CACHE_DIR
-tar czf $EXECUTOR.tgz $EXECUTOR
```
This will implicitly change the structure of the executor binary. Is that intentional?

LLParse replied:
Unintentional, I've reverted the change.

brndnmtthws commented:
A few notes:
 - Looks great! Thanks for the patch.
 - Unit tests would be a great addition.
 - Please include a description of how you tested/validate the change.

Commit:
Fix build

Commit:
revert executor structure

Commit:
Add loop for HA

Commit:
remove system property override, fix PMD hardcoded IP address warning

Commit:
remove process supervision (using while loop within bash script, this is dead code)

Commit:
Remove scaleUp stub

Commit:
data node size livestate test

Commit:
datanode to active transition test

elingg commented:
I left one response to your comment above. Please let me know when you are ready for me to take a second look, @LLParse, and I will review again ASAP.

Commit:
synchronize methods and logic related to current acquisition phase

Commit:
Only transition back to DATA_NODE phase if currently ACTIVE

Commit:
Rewrite scheduler config when data node count is mutated programmatically

Commit:
Make mesos-site.xml distributed to executor during config reloads, remove 'curl' dependency

Commit:
Add logging

LLParse commented:
@brndnmtthws I added some unit tests to cover the new phase and persistent state storage. In testing, I found a bug where submitting a scale-up request before the scheduler reaches ACTIVE state would cause things to break. I've committed a fix for that too. As for testing, I've run tests locally inside a virtual machine comprised of a 6-slave Mesos cluster (running inside of docker). Scale up requests appear to work fine and, additionally, using the dfsadmin tool to print the HDFS topology/status provided me with conclusive evidence that the added nodes are working correctly. Down the road, we might consider initiating a rebalance after scaleup (but I don't think that belongs in this PR), but to date that remains an exercise for the operator.

@elingg ready for another review. I've added code to update mesos-site.xml and things appear to be working well in the face of scheduler failover.

elingg commented:
Thanks @LLParse for your quick responses and excellent work, I'll take a look ASAP. A couple of things in the meantime:

1) Was this comment addressed? This is very important, because we need to make sure even if we don't have datanodes everywhere we still have the client binaries on all nodes. Otherwise, we launch Spark running on top of HDFS and Spark jobs start failing. ```A major issue here is that we still need to distribute the binaries on slaves where the datanodes are not running. To do this we need to launch a simple task with few resources, which distributes the binaries there. If not, people running say Spark on HDFS will encounter errors because they cannot fetch files on the slaves where the binaries are not distributed.```

2) We need to be sure to test all changes on Mesos, but also on DCOS before merging.

3) This is a minor change, but we need to update the README to let users know that they configure mesos-site.xml and hdfs-site.xml at runtime (doesn't need to be at compile time anymore) although that is an option.

Commit:
update readme

elingg commented on README.md:
```diff
@@ -25,7 +24,7 @@ Installing HDFS-Mesos on your Cluster
 1. Upload `hdfs-mesos-*.tgz` to a node in your Mesos cluster (which is built to `$PROJ_DIR/build/hdfs-mesos-x.x.x.tgz`).
 2. Extract it with `tar zxvf hdfs-mesos-*.tgz`.
-3. Optional: Customize any additional configurations that weren't updated at compile time in `hdfs-mesos-*/etc/hadoop/*-site.xml` Note that if you update hdfs-site.xml, it will be used by the scheduler and bundled with the executors. However, core-site.xml and mesos-site.xml will be used by the scheduler only.
+3. Configure `hdfs-site.xml` and `mesos-site.xml`. They will eventually be transmitted to each executor and override any configuration bundled with the executor at compile time. However, `core-site.xml` may only be configured at compile time.
```
Thanks for updating this.  This should still be marked as optional in case the user has already configured at compile time.

elingg commented on hdfs-commons/src/main/java/org/apache/mesos/hdfs/config/HdfsFrameworkConfig.java:
```diff
@@ -61,6 +66,25 @@ public HdfsFrameworkConfig() {
     setConf(configuration);
   }
 
+  public void writeXml() {
```
Let's separate this method into a separate class. Also, we need to think about concurrency here. What if there are multiple requests coming in at once. We need a form of locking.

elingg replied:
Maybe the best place for this is ConfigResource?

elingg commented on hdfs-executor/src/main/java/org/apache/mesos/hdfs/executor/AbstractNodeExecutor.java:
```diff
@@ -203,36 +204,35 @@ protected void startProcess(ExecutorDriver driver, Task task) {
-        log.error("Error reloading hdfs-site.xml.");
-      }
-    } catch (InterruptedException | IOException e) {
+      log.info(String.format("Reloading %s from %s", filename, configUri));
```
Nice clean up!

elingg commented on hdfs-scheduler/src/main/java/org/apache/mesos/hdfs/scheduler/HdfsScheduler.java:
```diff
@@ -369,13 +384,23 @@ private ExecutorInfo createExecutor(String taskIdName, String nodeName, String e
               CommandInfo.URI
                 .newBuilder()
-                .setValue(hdfsFrameworkConfig.getJreUrl())
+                .setValue(
```
Can we name this /api/mesosconfig?

elingg commented on hdfs-scheduler/src/main/java/org/apache/mesos/hdfs/scheduler/HdfsScheduler.java:
```diff
@@ -369,13 +384,23 @@ private ExecutorInfo createExecutor(String taskIdName, String nodeName, String e
               CommandInfo.URI
                 .newBuilder()
                 .setValue(
```
Can we name this api/hdfsconfig ?

elingg commented:
@LLParse. Did another review pass and left some more comments. Also, points 1) and 2) I made above still need to be addressed.

elingg commented:
Another issue I realized is with this change is that if the binaries are pre-distributed, our configuration does become inconsistent. This change requires more complicated logic than I thought, so maybe an easier alternative is to allow the user to configure the size each datanodes. Scale up would then involve adding more nodes to the cluster.

After we improve our configuration logic and also introduce storage drivers in Mesos, which we are working on (the ability to distribute hadoop binaries using Mesos), then this implementation may be simpler.

Please let me know if you would like to set up a meeting and discuss further.

crakjie commented:
This PR seem a really good idea why not pushing this up ?

triclambert commented:
Please note that this repository is now deprecated.  See [this issue](https://github.com/mesosphere/hdfs/issues/270#issuecomment-247233350) for a more detailed explanation.
